src/main/java/uk/ac/cam/cl/dtg/isaac/quiz/PgQuestionAttempts.java

```diff
@@ -78,7 +78,15 @@ public PgQuestionAttempts(final PostgresSqlDb ds, final ContentMapper objectMapp
     public void registerAnonymousQuestionAttempt(final String userId, final String questionPageId,
             final String fullQuestionId, final QuestionValidationResponse questionAttempt)
             throws SegueDatabaseException {
+
+        // Anonymous question attempts are stored in a JSONB object, need to atomically read-and-update this object:
         try (Connection conn = database.getDatabaseConnection()) {
+            conn.setAutoCommit(false);  // Start a transaction to hold a lock open during.
+            try (PreparedStatement pst = conn.prepareStatement("SELECT pg_advisory_xact_lock(?)")) {
+                pst.setLong(1, userId.hashCode());
+                pst.executeQuery();
+            }
```

Comment:
As grabbing a lock is blocking I think we should explicitly set a lock timeout: https://stackoverflow.com/questions/20963450/controlling-duration-of-postgresql-lock-waits/20963803#20963803